crates/core/src/table/file_format_options.rs

```diff
@@ -0,0 +1,308 @@
+    file_format_options: Option<FileFormatRef>,
+) -> Arc<dyn WriterPropertiesFactory> {
+    build_writer_properties_factory_ffo(file_format_options)
+        .unwrap_or_else(|| build_writer_properties_factory_default())
+}
+
+#[cfg(feature = "datafusion")]
+pub fn to_table_parquet_options_from_ffo(
+    file_format_options: Option<&FileFormatRef>,
+) -> Option<TableParquetOptions> {
+    file_format_options.map(|ffo| ffo.table_options().parquet)
+}
+
+#[cfg(feature = "datafusion")]
+pub fn state_with_file_format_options(
+    state: SessionState,
+    file_format_options: Option<&FileFormatRef>,
+) -> DeltaResult<SessionState> {
+    if let Some(ffo) = file_format_options {
+        ffo.update_session(&state)?;
+    }
+    // Convert file format options to parquet options and delegate to the common implementation
+    let parquet_options = to_table_parquet_options_from_ffo(file_format_options);
+    Ok(state_with_parquet_options(state, parquet_options.as_ref()))
+}
+
+#[cfg(feature = "datafusion")]
+pub fn state_with_parquet_options(
+    state: SessionState,
+    parquet_options: Option<&TableParquetOptions>,
+) -> SessionState {
+    if let Some(parquet) = parquet_options {
+        let mut sb = SessionStateBuilder::new_from_existing(state.clone());
+        let mut tbl_opts = TableOptions::new();
+        tbl_opts.parquet = parquet.clone();
+        tbl_opts.set_config_format(ConfigFileType::PARQUET);
+        sb = sb.with_table_options(tbl_opts);
+        let state = sb.build();
+        return state;
+    }
+    state
+}
+
+#[cfg(feature = "datafusion")]
+fn build_writer_properties_tpo(
+    table_parquet_options: &Option<TableParquetOptions>,
+) -> Option<WriterProperties> {
+    use datafusion::common::file_options::parquet_writer::ParquetWriterOptions;
+    table_parquet_options.as_ref().map(|tpo| {
+        let mut tpo = tpo.clone();
+        tpo.global.skip_arrow_metadata = true;
+        ParquetWriterOptions::try_from(&tpo)
+            .expect("Failed to convert TableParquetOptions to ParquetWriterOptions")
+            .writer_options()
+            .clone()
+    })
+}
+
+#[cfg(feature = "datafusion")]
+fn build_writer_properties_factory_tpo(
+    table_parquet_options: &Option<TableParquetOptions>,
+) -> Option<Arc<dyn WriterPropertiesFactory>> {
+    let props = build_writer_properties_tpo(table_parquet_options);
+    props.map(|wp| {
+        Arc::new(SimpleWriterPropertiesFactory::new(wp)) as Arc<dyn WriterPropertiesFactory>
+    })
+}
+
+pub fn build_writer_properties_factory_wp(
+    writer_properties: WriterProperties,
+) -> Arc<dyn WriterPropertiesFactory> {
+    Arc::new(SimpleWriterPropertiesFactory::new(writer_properties))
+}
+
+pub fn build_writer_properties_factory_default() -> Arc<dyn WriterPropertiesFactory> {
+    Arc::new(SimpleWriterPropertiesFactory::default())
+}
+
+#[async_trait]
+pub trait WriterPropertiesFactory: Send + Sync + std::fmt::Debug + 'static {
```

Comment:
I wonder if we could move these methods into `FileFormatOptions` rather than need a separate trait.

I guess this might be needed for backwards compatibility with setting writer properties though.